### PR TITLE
feat: add `array/base/count-same-value`

### LINKED ISSUE #1325: [RFC]: add `@stdlib/array/base/count-same-value`
### Description

This RFC proposes adding the package `@stdlib/array/base/count-same-value`.

The package should be structured similarly to [`@stdlib/array/base/count-truthy`](https://github.com/stdlib-js/stdlib/tree/64931c5fdb4abc3a339b619658d3b5acc089e556/lib/node_modules/%40stdlib/array/base/count-truthy).

The package should expose an API having the same signature as `countTruthy`, with the addition of a `value` argument.

```javascript
function countSameValue( x, value ) {...}
```

Internally, the package should use [`@stdlib/assert/is-same-value`](https://github.com/stdlib-js/stdlib/tree/64931c5fdb4abc3a339b619658d3b5acc089e556/lib/node_modules/%40stdlib/assert/is-same-value) to test whether an element `x[i]` is equal to a specified `value`.

Package: `@stdlib/array/base/count-same-value`
Alias: `countSameValue`

### Related Issues

Related issues #1324.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by skoriop**
Resolves #1325 .

## Description

> What is the purpose of this pull request?

This pull request adds `count-same-value` to `array/base`.

## Related Issues

> Does this pull request have any related issues?

This pull request resolves #1325 .

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/array/base/count-same-value/docs/repl.txt
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/docs/repl.txt:10]:
```suggestion

    value: any
```

File: lib/node_modules/@stdlib/array/base/count-same-value/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/README.md]:
```suggestion
var x = sample( [ 1, 2, 3, 4, 5 ] );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/docs/types/index.d.ts
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/docs/types/index.d.ts]:
```suggestion
* Counts the number of elements that are equal to a given value in an array.
```

File: lib/node_modules/@stdlib/array/base/count-same-value/examples/index.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/examples/index.js]:
```suggestion
var x = sample( [ 1, 2, 3, 4, 5 ] );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/README.md
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/README.md]:
```suggestion

var out = countSameValue( x, 1 );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/docs/repl.txt
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/docs/repl.txt]:
```suggestion
        Input value.
```

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
* Counts the number of elements that are equal to a given value in an indexed array.
```

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
		if ( isSameValue( x[ i ], value ) ) {
```

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
* Counts the number of elements that are equal to a given value in an accessor array.
```

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
		if ( isSameValue( get( x, i ), value ) ) {
```

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
* Counts the number of elements that are equal to a given value in a complex array.
```

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]: Hmm...this is a trickier one. (1) We probably want to avoid complex number object comparison, as it is slower. `x.get( i )`, will return a new complex number instance, and that construction will be relatively expensive. In which case, we'd like to avoid if possible. (2) We'd want to consider whether a real-valued `value` (e.g., `9.0`) should compare as equal to `9 + 0j` (i.e., a real-number having an implied zero for its imaginary component). Given that [`@stdlib/assert/is-same-value`](https://github.com/stdlib-js/stdlib/blob/64931c5fdb4abc3a339b619658d3b5acc089e556/lib/node_modules/%40stdlib/assert/is-same-value/lib/main.js) requires that both be complex number objects, I think the answer is no.

Nevertheless, we can improve this implementation in the following ways:

1. Before L111, we should check whether `value` is complex-like. If not, we should return `0`, as we know that only complex numbers can compare as equal.
2. If `value` is complex-like, we should decompose into real and imaginary components (e.g., using `@stdlib/complex/real`/`@stdlib/complex/realf`).
3. We should reinterpret the complex number array as a real-valued array using `@stdlib/strided/base/reinterpret-complex`.
4. The loop condition should be `i = 0; i < x.length; i += 2`
5. L113 should then be something along the lines of `if ( isSameValue( x[ i ], real ) && isSameValue( x[ i+1 ], imag ) {`.

This should avoid complex number instance creation and be much faster.

That make sense, @skoriop?

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
* Counts the number of elements that are equal to a given value in an array.
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js:48]: I think you want to also include test cases for where the same-value algorithm is applicable. For example, update `x` to include `NaN`, `-0`, and `+0` values and test for those.

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js:87]: Same comment as above.

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js:165]: Same comment as above. Would be useful to add test cases involving `NaN` and signed zero components.

**Review by kgryte — CHANGES_REQUESTED** [review]
@skoriop Thanks for working on this. Overall, this is looking pretty good. I left a suggestion for improving the complex array case and added suggestions for improving the test cases. Once resolved, this should be ready for merge.

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (skoriop) [reply lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]: Makes sense, done in 728628e

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (skoriop) [reply lib/node_modules/@stdlib/array/base/count-same-value/test/test.js:48]: Yup, done in 8a17c65

File: lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/lib/main.js]:
```suggestion
	if ( !isComplexLike( value ) ) {
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	x = [ 0, 1, 0, 1, 2 ];
```

For this test case, fine to leave off the negative sign.

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	x = [ 0.0, -0.0, 0.0, -0.0, 0.0, NaN, 1.0 ];
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	actual = countSameValue( x, 0.0 );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	x = toAccessorArray( [ 0.0, -0.0, 0.0, -0.0, 0.0, NaN, 1.0 ] );
	expected = 3;
	actual = countSameValue( x, 0.0 );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	x = new Float32Array( [ 0.0, -0.0, 0.0, -0.0, 0.0, NaN, 1.0 ] );
	expected = 3;
	actual = countSameValue( x, 0.0 );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	x = new Complex128Array( [ 0.0, -0.0, 0.0, -0.0, 0.0, 0.0, -0.0, -0.0, -0.0, 0.0 ] );
```

File: lib/node_modules/@stdlib/array/base/count-same-value/test/test.js
Comment (kgryte) [inline lib/node_modules/@stdlib/array/base/count-same-value/test/test.js]:
```suggestion
	x = new Complex128Array( [ NaN, NaN, NaN, 2.0, NaN, 9.0, NaN, NaN ] );
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @skoriop!

**Comment by kgryte** [comment]
I'll merge this once CI finishes.

**Pull request merged**
